From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Wed, 8 Nov 2023 21:43:53 -0700
Subject: [PATCH] Update .gitignore for Gradle build


diff --git a/.gitignore b/.gitignore
index 057f41d2039c55b862cb2051e0c14ea7e9e2c959..9210d06996180f942da21ca55daeb439d7a4c46a 100644
--- a/.gitignore
+++ b/.gitignore
@@ -8,3 +8,4 @@
 target/
 /bin
 /node_modules
+/config/build
