From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date: Sat, 11 Nov 2023 19:11:43 -0700
Subject: [PATCH] test


diff --git a/config/src/test/scala/com/typesafe/config/impl/TestUtils.scala b/config/src/test/scala/com/typesafe/config/impl/TestUtils.scala
index 353f17b67d969addc607af958e4a53978aedaa81..2787b9972e161cb12d0bc648024b7838f16b4c14 100644
--- a/config/src/test/scala/com/typesafe/config/impl/TestUtils.scala
+++ b/config/src/test/scala/com/typesafe/config/impl/TestUtils.scala
@@ -586,7 +586,7 @@ abstract trait TestUtils {
         val options = ConfigParseOptions.defaults().
             setOriginDescription("test string").
             setSyntax(ConfigSyntax.CONF);
-        ConfigFactory.parseString(s, options).asInstanceOf[SimpleConfig]
+        ConfigFactory.parseString(s.replace("\r\n", "\n"), options).asInstanceOf[SimpleConfig]
     }
 
     protected def subst(ref: String, optional: Boolean): ConfigReference = {
